### PR TITLE
Use new test framework in explorer to reduce per-test overhead

### PR DESCRIPTION
**Pull request by jonmeow**
Isolating test execution time (no build time included):

- Linux, `lit` test-per-file: Elapsed time: 28.214s, Critical Path: 13.97s
- Linux, `cc_test`-per-file: Elapsed time: 11.534s, Critical Path: 6.05s
- Linux, merged `cc_test` with 50 shards: Elapsed time: 11.677s, Critical Path: 11.17s
- Mac, `lit` test-per-file: Elapsed time: 295.686s, Critical Path: 20.00s
- Mac, `cc_test`-per-file: Elapsed time: 55.788s, Critical Path: 3.81s
- Mac, merged `cc_test` with 50 shards: Elapsed time: 16.269s, Critical Path: 7.54s

In GH actions:

- [Before](https://github.com/carbon-language/carbon-lang/actions/runs/4866602695/jobs/8678306144?pr=2799):
  - test / test (ubuntu-22.04, fastbuild) (pull_request_target) Successful in 20m
  - test / test (ubuntu-22.04, opt) (pull_request_target) Successful in 15m
  - test / test (macos-12, fastbuild) (pull_request_target) Successful in 36m
  - test / test (macos-12, opt) (pull_request_target) Successful in 21m
- [After](https://github.com/carbon-language/carbon-lang/actions/runs/4875154751/jobs/8697004066?pr=2811):
  - test / test (ubuntu-22.04, fastbuild) (pull_request_target) Successful in 10m
  - test / test (ubuntu-22.04, opt) (pull_request_target) Successful in 9m
  - test / test (macos-12, fastbuild) (pull_request_target) Successful in 12m
  - test / test (macos-12, opt) (pull_request_target) Successful in 9m

I'm still leaving a handful of `lit` tests to test end-to-end binary execution. This is why testdata directories are split (`lit` tests next to the `explorer` binary, the `cc_test`s next to `ParseAndExecute`).

### CODE REVIEW

File: explorer/parse_and_execute/file_test.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+extern auto RegisterFileTests(const std::vector<llvm::StringRef>& paths)
+    -> void {
+  ParseAndExecuteTestFile::RegisterTests(
+      "ParseAndExecuteTestFile", paths, [=](llvm::StringRef path) {
+        return new ParseAndExecuteTestFile(path, /*trace=*/true);
+      });
+  ParseAndExecuteTestFile::RegisterTests(
+      "ParseAndExecuteTestFile.debug", paths, [=](llvm::StringRef path) {
+        return new ParseAndExecuteTestFile(path, /*trace=*/false);
```
Comment (chandlerc) [inline explorer/parse_and_execute/file_test.cpp]: Above, the comment says only to capture trace streaming when in debug mode, but here the `.debug` suffix sets `trace` to `false`? Maybe I just don't understand the naming convention or flag behavior?

**Review by chandlerc — APPROVED** [review]
Generally LGTM, and not sure we should leave this open for too long given how disruptive it is. So I'd encourage landing this and we can fix anything forward.

Beyond the inline comment, I wonder whether it'd be more discoverable and less surprising to keep the bulk of the tests in `//explorer/testdata` and sink the lit-run tests into `//explorer/testdata/lit` or something similar? I mostly worry about it being a bit hard to discover the right way to add a test for the explorer with the current structure. WDYT?

Again, not a blocking change as making more changes with this many files doesn't seem good.

**Comment by jonmeow** [comment]
> Generally LGTM, and not sure we should leave this open for too long given how disruptive it is. So I'd encourage landing this and we can fix anything forward.
> 
> Beyond the inline comment, I wonder whether it'd be more discoverable and less surprising to keep the bulk of the tests in `//explorer/testdata` and sink the lit-run tests into `//explorer/testdata/lit` or something similar? I mostly worry about it being a bit hard to discover the right way to add a test for the explorer with the current structure. WDYT?
> 
> Again, not a blocking change as making more changes with this many files doesn't seem good.

Per discussion on #explorer, I've made this testdata and lit_testdata with lit_autoupdate.py and lit_autoupdate_lit.py to meet your request.

File: explorer/parse_and_execute/file_test.cpp (hunk 1 above)
Comment (jonmeow) [reply explorer/parse_and_execute/file_test.cpp]: Fixed, true/false were swapped.

**Review by chandlerc — APPROVED** [review]
LG, ship it!

**Pull request merged**